File: Java/Guias/guia3/ej7_BankAccount/BankAccount.java
package Guias.guia3.ej7_BankAccount;

public abstract class BankAccount {
    private int id;
    protected double balance;

    public BankAccount(int id) {
        this.id = id;
    }

    public void deposit(double amount) {
        this.balance += amount;
    }

    protected abstract boolean canExtract(double amount);

    public double extract(double amount) {
        if (canExtract(amount)) {
            this.balance -= amount;
            return this.balance;
        } else {
            System.out.println("Insufficient balance");
            return this.balance;
        }
    }

    public double getBalance(){
        return this.balance;
    }

    @Override
    public String toString() {
        return "Cuenta %d con saldo %.2f".formatted(this.id, this.balance);
    }
}

File: RUBY/Guias/Guia8/Ej5_CheckingAccount/bank_account.rb
# frozen_string_literal: true

class BankAccount

  def initialize
    raise 'Debe sobreescribir este método'
  end

  def create(id)
    @id = id
    @saldo = 0
  end

  def deposit(amount)
    @saldo += amount
  end

  def canExtract?(amount)
    raise 'Se debe sobreescribir'
  end

  def extract(amount)
    raise 'No cuenta con los fondos necesarios' unless canExtract? amount
    @saldo -= amount
  end

  def to_s
    "Cuenta #{@id} con saldo #{@saldo}"
  end

  private :create, :canExtract?
end

